File: test/sync_scope_test.dart
import 'package:aichat/models/sync_policy.dart';
import 'package:aichat/services/sync_scope.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('SyncScope', () {
    test('all mode includes future agents and all legacy tables', () {
      final scope = SyncScope.accountPolicy(const SyncPolicy.all(version: 1));

      expect(scope.allowsTable('providers'), isTrue);
      expect(scope.allowsAgent('future-agent'), isTrue);
      expect(scope.tables, hasLength(13));
    });

    test('selected mode keeps only the private six-table closure', () {
      final scope = SyncScope.accountPolicy(
        const SyncPolicy.selected(agentIds: {'agent-a'}, version: 2),
      );

      expect(scope.tables, agentSyncClosureTables);
      expect(
        scope.filterRows('agents', [
          {'client_id': 'agent-a'},
          {'client_id': 'agent-b'},
        ]),
        [
          {'client_id': 'agent-a'},
        ],
      );
      expect(
        scope.filterRows('long_term_memories', [
          {'client_id': 'm1', 'agent_id': 'agent-a', 'group_id': null},
          {'client_id': 'm2', 'agent_id': 'agent-b', 'group_id': null},
          {'client_id': 'm3', 'agent_id': 'agent-a', 'group_id': 'group-1'},
        ]),
        [
          {'client_id': 'm1', 'agent_id': 'agent-a', 'group_id': null},
        ],
      );
    });

    test('one-shot scope does not mutate account policy', () {
      const policy = SyncPolicy.selected(
        agentIds: {'agent-a', 'agent-b'},
        version: 3,
        realtimeEnabled: true,
      );

      final oneShot = SyncScope.oneShot({'agent-c'});

      expect(oneShot.agentIds, {'agent-c'});
      expect(policy.selectedAgentIds, {'agent-a', 'agent-b'});
      expect(policy.realtimeEnabled, isTrue);
    });

    test('selected tombstones require an in-scope agent', () {
      final scope = SyncScope.oneShot({'agent-a'});
      final filtered = scope.filterTombstones([
        {'table_name': 'agents', 'client_id': 'agent-a'},
        {'table_name': 'agents', 'client_id': 'agent-b'},
        {
          'table_name': 'chat_messages',
          'client_id': 'message-a',
          'agent_id': 'agent-a',
        },
        {
          'table_name': 'chat_messages',
          'client_id': 'message-b',
          'agent_id': 'agent-b',
        },
      ]);

      expect(filtered.map((item) => item['client_id']), [
        'agent-a',
        'message-a',
      ]);
    });
  });
}

File: website/API/services/sync_scope_test.go
package services

import "testing"

func TestSelectedSyncScopeKeepsOnlyPrivateAgentClosure(t *testing.T) {
	scope := NewSelectedSyncScope([]string{"agent-a"})
	items := []map[string]interface{}{
		{"client_id": "m1", "agent_id": "agent-a", "group_id": ""},
		{"client_id": "m2", "agent_id": "agent-b", "group_id": ""},
		{"client_id": "m3", "agent_id": "agent-a", "group_id": "group-1"},
	}

	filtered := FilterSyncItems(scope, "long_term_memories", items)
	if len(filtered) != 1 || filtered[0]["client_id"] != "m1" {
		t.Fatalf("filtered = %#v, want only m1", filtered)
	}
}

func TestSelectedSyncScopeRejectsGlobalAndGroupTables(t *testing.T) {
	scope := NewSelectedSyncScope([]string{"agent-a"})
	for _, table := range []string{"group_chats", "group_messages", "user_profiles", "providers"} {
		if got := FilterSyncItems(scope, table, []map[string]interface{}{{"client_id": "x"}}); len(got) != 0 {
			t.Fatalf("table %s returned %#v, want empty", table, got)
		}
	}
}

func TestAllSyncScopeIncludesFutureAgentsAndGlobalTables(t *testing.T) {
	scope := NewAllSyncScope()
	items := []map[string]interface{}{{"client_id": "new-agent"}}
	if got := FilterSyncItems(scope, "agents", items); len(got) != 1 {
		t.Fatalf("agents = %#v, want new agent", got)
	}
	if got := FilterSyncItems(scope, "providers", items); len(got) != 1 {
		t.Fatalf("providers = %#v, want global table in all mode", got)
	}
}

func TestSelectedSyncScopeFiltersTombstonesByAgent(t *testing.T) {
	scope := NewSelectedSyncScope([]string{"agent-a"})
	tombstones := []map[string]interface{}{
		{"table_name": "agents", "client_id": "agent-a"},
		{"table_name": "agents", "client_id": "agent-b"},
		{"table_name": "chat_messages", "client_id": "c1", "agent_id": "agent-a"},
		{"table_name": "chat_messages", "client_id": "c2", "agent_id": "agent-b"},
	}

	filtered := FilterSyncTombstones(scope, tombstones)
	if len(filtered) != 2 {
		t.Fatalf("filtered tombstones = %#v, want 2", filtered)
	}
}
